### PR TITLE
tests: add impersonation to integration test

### LINKED ISSUE #2617: Integration Test ServiceAccount Impersonation
### Problem Statement

Historically in the integration tests we've used whatever `kubeconfig` was available to run the controller manager for the integration tests. Generally speaking this has meant we've run the integration tests with `cluster-admin` permissions. While we do test our explicit RBAC permissions in the E2E tests and this has otherwise caught permissions problems in the past, it would be an improvement to use these permissions for the integration tests as well for faster feedback.

### Additional information

The [Gateway Operator](https://github.com/kong/gateway-operator) has an example implementation of this, it uses impersonation to run the integration tests using the RBAC permissions present in `config/` in the test suite.

### Acceptance Criteria

- [ ] as a contributor if I make changes that require or include RBAC changes the integration tests will utilize those permissions so that I can get fast feedback if there's a problem.

### PR DESCRIPTION
**Pull request by parkjeongryul**
**What this PR does / why we need it**:

- Use RBACs permissions in integration test for faster feedback related permission problems.
- We've run the integration tests with cluster-admin permissions.

**Which issue this PR fixes**:

fixes #2617 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/util/test/rbacs.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+const (
+	kongRBACsKustomize        = "../../config/rbac/"
+	kongKnativeRBACsKustomize = "../../config/rbac/knative"
+	kongGatewayRBACsKustomize = "../../config/rbac/gateway"
+	kongCRDsRBACsKustomize    = "../../config/rbac/crds"
+)
+
+func DeployRBACsForCluster(ctx context.Context, cluster clusters.Cluster) error {
+	// create a tempfile to hold the cluster kubeconfig that will be used for the controller
+	// generate a temporary kubeconfig since we're going to be using the helm CLI
```
Comment (czeslavo) [inline internal/util/test/rbacs.go:23]: Nit: we do not use helm here.

File: internal/util/test/rbacs.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
+	args := []string{"--kubeconfig", kubeconfig.Name(), "kustomize", kongRBACsKustomize}
+	cmd := exec.CommandContext(ctx, "kubectl", args...)
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to deploy kong RBACs STDOUT=(%s) STDERR=(%s): %w", strings.TrimSpace(stdout.String()), strings.TrimSpace(stderr.String()), err)
+	}
+	fmt.Printf("INFO: running kubectl kustomize for Kong RBACs (args: %v)\n", args)
+	kongRBACsYAML := stdout.String()
```
Comment (czeslavo) [inline internal/util/test/rbacs.go:40]: I think we could use `clusters.KustomizeDeployForCluster` here and avoid the code duplication we have right now.

**Review by czeslavo — COMMENTED** [review]
Thank you for the PR! It looks alright, just a couple of nits to fix from my side.

**Comment by parkjeongryul** [comment]
updated:

I think that `clusters.KustomizeDeployForCluster` is appropriate function for this.
And i think that we need to replace crds.go with this function for consistency.

**Comment by parkjeongryul** [comment]
@czeslavo 

I ran the integration test in my local and got the following error. 
```bash
=== RUN   TestTCPRouteExample
    examples_test.go:154: locking TCP port
    examples_test.go:165: performing test setup
    examples_test.go:165: creating a testing namespace
    examples_test.go:165: created namespace 85663534-d0df-4aac-b858-2fb2deee74be for test case TestTCPRouteExample
    examples_test.go:167: applying yaml manifest ../../examples/gateway-tcproute.yaml
    examples_test.go:173: verifying that TCPRoute becomes routable
E0724 15:34:31.000976  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:34:34.132071  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:34:36.794202  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:34:41.098232  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:34:45.475963  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
...
E0724 15:34:49.731856  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:34:52.043117  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:34:55.789159  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:34:59.100314  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:35:02.851445  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:24.864153  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:27.050934  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:29.862488  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
    examples_test.go:174:
                Error Trace:    /home1/irteamsu/kubernetes-ingress-controller/test/integration/examples_test.go:174
                Error:          Condition never satisfied
                Test:           TestTCPRouteExample
E0724 15:37:32.685467  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:36.758391  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:39.961223  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:42.574488  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:45.847833  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:49.749020  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:53.604852  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
    setup.go:33: TestTCPRouteExample failed, dumped diagnostics to /tmp/ktf-diag-2978671423
E0724 15:37:56.699788  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
E0724 15:37:59.829061  113924 leaderelection.go:327] error retrieving resource lock kong-system/integrationtests.konghq.com: leases.coordination.k8s.io "integrationtests.konghq.com" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "kong-system"
    examples_test.go:157: unlocking TCP port
--- FAIL: TestTCPRouteExample (211.00s)
```

In my opinion, the service-account exists in the kong namespace and the lease resource exists in the kong-system namespace. 

https://github.com/Kong/kubernetes-ingress-controller/blob/08c67a9ba99559c3bec9ebcac8a95eb0204887d1/config/rbac/leader_election_role.yaml#L2-L7

However, the corresponding RBAC is Role instead of **ClusterRole**. Is it a good idea to change this to ClusterRole?

**Comment by rainest** [comment]
It should remain a Role; that's what it should be in an actual install. We don't want access to ConfigMaps or Leases other than our own.

The default namespace of the manifests is `kong`, whereas the test namespace is `kong-system`. You should see the other namespace and role if you check for it. Are you able to copy that role/rolebinding into kong-system for the tests?

**Comment by parkjeongryul** [comment]
@rainest 
Thanks for pointing out the difference between the default namespace and the namespace used in the test. 

Updated: deploy to the test namespace (kong-system).
And i found that the failed test passed.
```bash
--- PASS: TestTCPRouteExample (14.09s)
```

**Review by czeslavo — APPROVED** [review]
I think we can merge it with the explicit `-n kong` installation of RBACs and tackle the possible namespace unification in https://github.com/Kong/kubernetes-ingress-controller/issues/4431.

Thank you for the contribution @parkjeongryul!

**Comment by czeslavo** [comment]
@parkjeongryul There's one linter issue left, could you please fix it?

**Comment by parkjeongryul** [comment]
> There's one linter issue left, could you please fix it?

Fixed. @czeslavo 
Thank you for your review.

**Comment by czeslavo** [comment]
@parkjeongryul There's one more linting issue (wasn't detected before as checks run serially).

**Comment by parkjeongryul** [comment]
> There's one more linting issue (wasn't detected before as checks run serially).



Fixed it. https://github.com/Kong/kubernetes-ingress-controller/pull/4376/commits/b6718d9732f0e50644439df158fb91f7c55fbe96

**Review by czeslavo — APPROVED** [review]

**Comment by parkjeongryul** [comment]
> I believe you can get a KIC contributor T-shirt by filling this [form](https://docs.google.com/forms/d/e/1FAIpQLSfacQHac-PYIDIrDOi_W2l7cWxBDRIVBs6xE5HnlojPvdhL9g/viewform). 🙂

Wow, thank you so much! I wasn't expecting this at all. 😃

**Pull request merged**